pkg/dyninst/procmon/analyze.go

```diff
@@ -296,3 +326,121 @@ func isGoElfBinaryWithDDTraceGo(f *os.File) (bool, error) {
 }
 
 var ddTraceSymbolSuffix = []byte("ddtrace/tracer.passProbeConfiguration")
+
+// executableAnalyzer is an interface for analyzing executables to determine if
+// they are interesting for dynamic instrumentation.
+type executableAnalyzer interface {
+	checkFileKeyCache(key FileKey) (interesting bool, known bool)
+	isInteresting(f *os.File, key FileKey) (bool, error)
+}
+
+// baseExecutableAnalyzer implements executableAnalyzer without any caching.
+type baseExecutableAnalyzer struct{}
+
+func (a *baseExecutableAnalyzer) checkFileKeyCache(
+	_ FileKey,
+) (interesting bool, known bool) {
+	return false, false
+}
+
+func (a *baseExecutableAnalyzer) isInteresting(
+	f *os.File, _ FileKey,
+) (bool, error) {
+	return isGoElfBinaryWithDDTraceGo(f)
+}
+
+type fileKeyCacheExecutableAnalyzer struct {
+	inner        executableAnalyzer
+	fileKeyCache *lru.Cache[FileKey, bool]
+}
+
+func newFileKeyCacheExecutableAnalyzer(
+	cacheSize int,
+	inner executableAnalyzer,
+) executableAnalyzer {
+	return &fileKeyCacheExecutableAnalyzer{
+		inner:        inner,
+		fileKeyCache: mustNewLruCache[FileKey, bool](cacheSize),
+	}
+}
+
+func (a *fileKeyCacheExecutableAnalyzer) checkFileKeyCache(
+	key FileKey,
+) (interesting bool, known bool) {
+	return a.fileKeyCache.Get(key)
+}
+
+func (a *fileKeyCacheExecutableAnalyzer) isInteresting(
+	f *os.File,
+	key FileKey,
+) (bool, error) {
+	if interesting, ok := a.fileKeyCache.Get(key); ok {
+		return interesting, nil
+	}
+	interesting, err := a.inner.isInteresting(f, key)
+	if err != nil {
+		return false, err
+	}
+	a.fileKeyCache.Add(key, interesting)
+	return interesting, nil
+}
+
+type htlHashCacheExecutableAnalyzer struct {
+	inner        executableAnalyzer
+	htlHashCache *lru.Cache[string, bool]
+}
+
+func newHtlHashCacheExecutableAnalyzer(
+	cacheSize int,
+	inner executableAnalyzer,
+) executableAnalyzer {
+	return &htlHashCacheExecutableAnalyzer{
+		inner:        inner,
+		htlHashCache: mustNewLruCache[string, bool](cacheSize),
+	}
+}
+
+func (a *htlHashCacheExecutableAnalyzer) isInteresting(
+	f *os.File,
+	key FileKey,
+) (bool, error) {
+	hash, err := computeHtlHash(f)
+	if err != nil {
+		return false, err
+	}
+	if interesting, ok := a.htlHashCache.Get(hash); ok {
+		return interesting, nil
+	}
+	interesting, err := a.inner.isInteresting(f, key)
+	if err != nil {
+		return false, err
+	}
+	a.htlHashCache.Add(hash, interesting)
+	return interesting, nil
+}
+
+func (a *htlHashCacheExecutableAnalyzer) checkFileKeyCache(
+	key FileKey,
+) (interesting bool, known bool) {
+	return a.inner.checkFileKeyCache(key)
+}
+
+func makeExecutableAnalyzer(cacheSize int) executableAnalyzer {
+	if cacheSize <= 0 {
+		return &baseExecutableAnalyzer{}
+	}
+	return newFileKeyCacheExecutableAnalyzer(
+		cacheSize,
+		newHtlHashCacheExecutableAnalyzer(cacheSize, &baseExecutableAnalyzer{}),
+	)
+}
+
+// mustNewLruCache panics if the cache creation fails which will only happen
+// if the size is non-positive.
```

Comment:
What's the advantage of having a panic-y constructor like this? Why not possibly return an error, or take a uint instead?